### ISSUE TITLE
Investigate what we can do to prevent `required` misalignment

### ISSUE THREAD

**Issue by untitaker** [getsentry/sentry-kafka-schemas#148]
(aka SNS-2325)

From [jsonschema-gentypes issue](https://github.com/sbrunner/jsonschema-gentypes/issues/483):

We have a schema definition with two fields, foo_required and foo_optional.

```
{"properties": {"foo_required": {}, "foo_optional": {}}, "required": ["foo_required"]}
```

Now consider this code:

```
x["foo_optional"]
```

VSCode/pyright does error on the above line because it might crash. Unfortunately, mypy does not.

We had an incident (without incident ticket) where this exact crash happened:
https://github.com/getsentry/sentry-kafka-schemas/commit/c19ad275cfac47cd1109e72daab6c68cec11d869 

This schema change added a new, optional field, but the corresponding Snuba PR accessed the field as if it was required. Mypy did not catch this.

When we merged the Snuba PR, sentry CI failed because the field wasn’t being sent to snuba, so snuba crashed. Technically there were multiple problems:

1) schema validation was missing in some places on producer side

2) even if the schema was validated on producer side, the field was optional according to it

3) on the snuba side, the misalignment between consumer and schema was not caught due to the above mypy bug

We should focus on (3) for this ticket. If the field was marked as required in the schema, our existing PR linting would’ve at least told the user that this was a breaking change.
